# Common table expressions

Using the northwind database

## 1. Write a CTE that lists the names and quantities of products with a unit price greater than $50.

In [8]:
import pymysql 
import sqlalchemy as alch 
from getpass import getpass
import pandas as pd
import os

In [9]:
password = getpass()
dbName = "northwind"
connectionData=f"mysql+pymysql://root:{password}@localhost/{dbName}"
engine = alch.create_engine(connectionData) 

········


In [10]:
query = "SELECT * FROM products;"
pd.read_sql_query(query, engine)

ProductID                      ProductName  SupplierID  CategoryID  \
0           1                            Chais           1           1   
1           2                            Chang           1           1   
2           3                    Aniseed Syrup           1           2   
3           4     Chef Anton's Cajun Seasoning           2           2   
4           5           Chef Anton's Gumbo Mix           2           2   
..        ...                              ...         ...         ...   
72         73                       Röd Kaviar          17           8   
73         74                    Longlife Tofu           4           7   
74         75             Rhönbräu Klosterbier          12           1   
75         76                     Lakkalikööri          23           1   
76         77  Original Frankfurter grüne Soße          12           2   

                   Unit  Price  
0    10 boxes x 20 bags   18.0  
1    24 - 12 oz bottles   19.0  
2   12 - 550 ml bottles   10.0  
3        48 - 6 oz jars   22.0  
4              36 boxes   21.0  
..                  ...    ...  
72      24 - 150 g jars   15.0  
73            5 kg pkg.   10.0  
74   24 - 0.5 l bottles    8.0  
75               500 ml   18.0  
76             12 boxes   13.0  

[77 rows x 6 columns]

In [11]:
query = ''' SELECT productname, unit
FROM products
WHERE price > 50;
'''
pd.read_sql_query(query, engine)

productname                  unit
0          Mishi Kobe Niku      18 - 500 g pkgs.
1         Carnarvon Tigers            16 kg pkg.
2   Sir Rodney's Marmalade         30 gift boxes
3  Thüringer Rostbratwurst  50 bags x 30 sausgs.
4            Côte de Blaye    12 - 75 cl bottles
5    Manjimup Dried Apples      50 - 300 g pkgs.
6     Raclette Courdavault             5 kg pkg.

#### Expected results

|ProductName                     |Unit|
|--------------------------------|----|
|Mishi Kobe Niku                 |18 - 500 g pkgs.|
|Carnarvon Tigers                |16 kg pkg.|
|Sir Rodney's Marmalade          |30 gift boxes|
|Thüringer Rostbratwurst         |50 bags x 30 sausgs.|
|Côte de Blaye                   |12 - 75 cl bottles|
|Manjimup Dried Apples           |50 - 300 g pkgs.|
|Raclette Courdavault            |5 kg pkg.|


## 2. What are the top 5 most profitable products?

In [12]:
query = '''SELECT products.productname AS productname,
        SUM(orderdetails.quantity * products.price) AS totalprofit
    FROM
        products
    JOIN
        orderdetails ON products.productid = orderdetails.productid
    GROUP BY
        products.productid
    ORDER BY
        totalprofit DESC
    LIMIT 5;
'''
pd.read_sql_query(query, engine)

productname  totalprofit
0            Côte de Blaye      63096.0
1  Thüringer Rostbratwurst      20832.0
2     Raclette Courdavault      19030.0
3           Tarte au sucre      15925.0
4        Camembert Pierrot      14620.0

#### Expected results
Alice Mutton is 6th

|ProductID                       |ProductName|TotalRevenue|
|--------------------------------|-----------|------------|
|-                             |-|-       |
|-                           |-|-       |
|-                            |-|-       |
|-                             |-|-       |
|-                              |-|-       |
|17                              |Alice Mutton|12909       |

## 3. Write a CTE that lists the top 5 categories by the number of products they have.

In [13]:
query = '''SELECT
    categories.categoryname AS categoryname,
    COUNT(products.productid) AS productcount
FROM
    categories
LEFT JOIN
    products ON categories.categoryid = products.categoryid
GROUP BY
    categories.categoryname
ORDER BY
    productcount DESC
LIMIT 5;
'''
pd.read_sql_query(query, engine)

categoryname  productcount
0     Confections            13
1       Beverages            12
2      Condiments            12
3         Seafood            12
4  Dairy Products            10

#### Expected result
|CategoryName                    |ProductCount|
|--------------------------------|------------|
|Confections                     |13          |
|Beverages                       |12          |
|Condiments                      |12          |
|Seafood                         |12          |
|Dairy Products                  |10          |


## 4. Write a CTE that shows the average order quantity for each product category.

In [14]:
query = '''SELECT
    categories.categoryname AS categoryname,
    AVG((SELECT AVG(orderdetails.quantity) FROM orderdetails WHERE orderdetails.productid = products.productid)) AS averageorder
FROM
    categories 
JOIN
    products ON categories.categoryid = products.categoryid
GROUP BY
    CategoryName
ORDER BY
    CategoryName;
'''
pd.read_sql_query(query, engine)

categoryname  averageorder
0       Beverages     23.160958
1      Condiments     32.960642
2     Confections     28.486692
3  Dairy Products     24.296040
4  Grains/Cereals     20.908157
5    Meat/Poultry     25.198483
6         Produce     20.827620
7         Seafood     21.354300

|CategoryName                    |AvgOrderQuantity|
|--------------------------------|----------------|
|Beverages                       |24.6129         |
|Condiments                      |28.2245         |
|Confections                     |25.1190         |
|Dairy Products                  |26.0100         |
|Grains/Cereals                  |21.7143         |
|Meat/Poultry                    |25.7600         |
|Produce                         |21.6667         |
|Seafood                         |21.5672         |


# 5. Create a CTE to calculate the average order amount for each customer.

In [16]:
query = '''
WITH AverageOrderAmount AS (
    SELECT
        c.CustomerID,
        c.CustomerName,
        AVG(od.Quantity * p.Price) AS AvgOrderAmount
    FROM
        Customers c
    JOIN
        Orders o ON c.CustomerID = o.CustomerID
    JOIN
        OrderDetails od ON o.OrderID = od.OrderID
	JOIN
		Products p ON p.ProductId = od.ProductID
    GROUP BY
        c.CustomerID, c.CustomerName
)
SELECT *
FROM AverageOrderAmount 
ORDER BY AvgOrderAmount DESC;


'''
pd.read_sql_query(query, engine)

CustomerID                        CustomerName  AvgOrderAmount
0           59                    Piccolo und mehr       4014.2500
1           73                       Simons bistro       2444.3333
2           62                       Queen Cozinha       1991.6667
3           51                      Mère Paillarde       1673.8571
4           71                  Save-a-lot Markets       1407.2500
..         ...                                 ...             ...
69          29              Galería del gastrónomo        147.7500
70          69                    Romero y tomillo        116.5000
71          13          Centro comercial Moctezuma         63.0000
72          27                      Franchi S.p.A.         62.0000
73           2  Ana Trujillo Emparedados y helados         55.5000

[74 rows x 3 columns]

|CustomerID                      |CustomerName|AvgOrderAmount|
|--------------------------------|------------|--------------|
|59                              |Piccolo und mehr|4014.2500     |
|73                              |Simons bistro|2444.3333     |
|62                              |Queen Cozinha|1991.6667     |
|51                              |Mère Paillarde|1673.8571     |
|71                              |Save-a-lot Markets|1407.2500     |
|76                              |Suprêmes délices|1345.8333     |
|81                              |Tradição Hipermercados|1315.6667     |
|7                               |Blondel père et fils|1174.4615     |
|89                              |White Clover Markets|1112.5000     |
|55                              |Old World Delicatessen|1079.5000     |
|20                              |Ernst Handel|1018.0000     |
|19                              |Eastern Connection|1004.8000     |
|68                              |Richter Supermarkt|976.6667      |
|72                              |Seven Seas Imports|934.0000      |
|63                              |QUICK-Stop  |908.5500      |
|25                              |Frankenversand|895.0667      |
|75                              |Split Rail Beer & Ale|854.7692      |
|65                              |Rattlesnake Canyon Grocery|838.2727      |
|52                              |Morgenstern Gesundkost|754.0000      |
|9                               |Bon app''   |750.4286      |
|37                              |Hungry Owl All-Night Grocers|733.5714      |
|35                              |HILARIÓN-Abastos|722.8333      |
|33                              |GROSELLA-Restaurante|690.0000      |
|34                              |Hanari Carnes|681.0000      |
|15                              |Comércio Mineiro|677.5000      |
|23                              |Folies gourmandes|672.3333      |
|10                              |Bottom-Dollar Marketse|664.0000      |
|31                              |Gourmet Lanchonetes|641.0000      |
|60                              |Princesa Isabel Vinhoss|628.5000      |
|5                               |Berglunds snabbköp|601.5556      |
|46                              |LILA-Supermercado|564.7692      |
|44                              |Lehmanns Marktstand|545.7500      |
|30                              |Godos Cocina Típica|515.0000      |
|3                               |Antonio Moreno Taquería|504.0000      |
|88                              |Wellington Importadora|503.8333      |
|47                              |LINO-Delicateses|500.0000      |
|56                              |Ottilies Käseladen|500.0000      |
|87                              |Wartian Herkku|492.5833      |
|24                              |Folk och fä HB|480.2222      |
|49                              |Magazzini Alimentari Riuniti|464.2857      |
|86                              |Die Wandernde Kuh|441.9091      |
|41                              |La maison d''Asie|437.8182      |
|8                               |Bólido Comidas preparadas|416.0000      |
|83                              |Vaffeljernet|411.0000      |
|14                              |Chop-suey Chinese|374.8333      |
|80                              |Tortuga Restaurante|357.4167      |


# 6. Sales Analysis with CTEs

Assume we have the Northwind database which contains tables like Orders, OrderDetails, and Products. Create a CTE that calculates the total sales for each product in the year 1997.

In [17]:
query = '''SELECT products.productname, SUM(orderdetails.quantity) AS totalsales
FROM orderdetails 
JOIN orders ON orders.orderid = orderdetails.orderid
JOIN products ON products.productid = orderdetails.productid
WHERE YEAR(orders.orderdate) = 1997
GROUP BY products.productname
ORDER BY totalsales DESC
LIMIT 15;
 '''
pd.read_sql_query(query, engine)

productname  totalsales
0   Gnocchi di nonna Alice       173.0
1                Tourtière       126.0
2                  Geitost       119.0
3     Raclette Courdavault       115.0
4                    Chang       115.0
5           Sirop d'érable       106.0
6             Vegie-spread       100.0
7            Côte de Blaye        99.0
8             Alice Mutton        97.0
9           Steeleye Stout        95.0
10     Sir Rodney's Scones        92.0
11                 Pavlova        86.0
12           Zaanse koeken        85.0
13          Tarte au sucre        75.0
14             Fløtemysost        75.0

#### Expected result

|ProductName                     |TotalSales|
|--------------------------------|----------|
|Gnocchi di nonna Alice          |173       |
|Tourtière                       |126       |
|Geitost                         |119       |
|Chang                           |115       |
|Raclette Courdavault            |115       |
|Sirop d'érable                  |106       |
|Vegie-spread                    |100       |
|Côte de Blaye                   |99        |
|Alice Mutton                    |97        |
|Steeleye Stout                  |95        |
|Sir Rodney's Scones             |92        |
|Pavlova                         |86        |
|Zaanse koeken                   |85        |
|Fløtemysost                     |75        |
|Tarte au sucre                  |75        |
